kvs/src/client/python/anna/client.py

```diff
@@ -0,0 +1,195 @@
+
+        self.elb_addr = elb_addr
+        self.elb_ports = elb_ports
+
+        if ip:
+            self.ut = UserThread(ip, offset)
+        else:
+            self.ut = UserThread(socket.gethostbyname(socket.gethostname()), offset)
+
+        self.context = zmq.Context(1)
+
+        self.address_cache = {}
+        self.pusher_cache = SocketCache(self.context, zmq.PUSH)
+
+        self.response_puller = self.context.socket(zmq.PULL)
+        self.response_puller.bind(self.ut.get_request_pull_bind_addr())
+
+        self.key_address_puller = self.context.socket(zmq.PULL)
+        self.key_address_puller.bind(self.ut.get_key_address_bind_addr())
+
+        self.rid = 0
+
+
+    def get(self, key):
+        worker_address = self._get_worker_address(key)
+        send_sock = self.pusher_cache.get(worker_address)
+
+        req, _ = self._prepare_data_request(key)
+        req.type = GET
+
+        send_request(req, send_sock)
+        response = recv_response([req.request_id], KeyResponse,
+                self.response_puller)[0]
+
+        # we currently only support single key operations
+        tup = resp_obj.tuples[0]
+
+        if tup.invalidate:
+            self._invalidate_cache(tup.key, tup.addresses)
+
+            # re-issue the request
+            return self.get(tup.key)
+
+        return str(tup.value, 'utf-8')
+
+    def get_all(self, key):
+        worker_addresses = self._get_worker_address(key, False)
+
+        req, _ = self._prepare_data_request(key)
+        req.type = GET
+
+        req_ids = []
+        for address in worker_addresses:
+            # NOTE: We technically waste a request id here, but it doesn't
+            # really matter
+            req.request_id = self._get_request_id()
+
+            send_sock = self.pusher_cache.get(address)
+            send_request(req, send_sock)
+
+            req_ids.append(req.request_id)
+
+        responses = recv_response(req_ids, self.response_puller, KeyResponse)
+
+        return list(map(lambda resp: str(resp.tuples[0].value, 'utf-8'), responses))
+
+
+    def durable_put(self, key, value):
+        worker_addresses = self._get_worker_address(key, False)
+
+        req, tup = self._prepare_data_request(key)
+        req.type = PUT
+        tup.value = bytes(value, 'utf-8')
+        tup.timestamp = 0
+
+        req_ids = []
+        for address in worker_addresses:
+            # NOTE: We technically waste a request id here, but it doesn't
+            # really matter
+            req.request_id = self._get_request_id()
+
+            send_sock = self.pusher_cache.get(address)
+            send_request(req, send_sock)
+
+            req_ids.append(req.request_id)
+
+        responses = recv_response(req_ids, self.response_puller, KeyResponse)
+
+        for resp in responses:
+            tup = resp.tuples[0]
+            if tup.invalidate:
+                # reissue the request
+                return self.durable_put(key, value)
+
+            if tup.error != 0:
```

Comment:
Just to confirm: if invalidation is false, then in principle `tup.error` should be 0 right? So if `tup.error != 0` something must be wrong so we are just returning false?

Reply:
Yes, `False` just means that the operation failed, and you should try again.